observability/logging.go

```diff
@@ -0,0 +1,232 @@
+		grpcLogRecord.Message = binlogEntry.GetMessage().GetData()
+		grpcLogRecord.PayloadSize = binlogEntry.GetMessage().GetLength()
+		grpcLogRecord.PayloadTruncated = binlogEntry.GetPayloadTruncated()
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_CLIENT_HALF_CLOSE:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_HALF_CLOSE
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_SERVER_TRAILER:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_TRAILER
+		grpcLogRecord.Metadata = translateMetadata(binlogEntry.GetTrailer().Metadata)
+		grpcLogRecord.StatusCode = binlogEntry.GetTrailer().GetStatusCode()
+		grpcLogRecord.StatusMessage = binlogEntry.GetTrailer().GetStatusMessage()
+		grpcLogRecord.StatusDetails = binlogEntry.GetTrailer().GetStatusDetails()
+		grpcLogRecord.PayloadTruncated = binlogEntry.GetPayloadTruncated()
+		setPeerIfPresent(binlogEntry, grpcLogRecord)
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_CANCEL:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_CANCEL
+	default:
+		logger.Warningf("unknown event type: %v", binlogEntry.Type)
+		return
+	}
+	grpcLogRecord.ServiceName = ml.serviceName
+	grpcLogRecord.MethodName = ml.methodName
+	// CloudLogging client doesn't return error on entry write. Entry writes
+	// don't mean the data will be uploaded immediately.
+	globalLoggingExporter.EmitGrpcLogRecord(grpcLogRecord)
+}
+
+type observabilityBinaryLogger struct {
+	// originalLogger is needed to ensure binary logging users won't be impacted
+	// by this plugin. Users are allowed to subscribe to a completely different
+	// set of methods.
+	originalLogger iblog.Logger
+	// wrappedLogger is needed to reuse the control string parsing, logger
+	// config managing logic
+	wrappedLogger iblog.Logger
+}
+
+func (l *observabilityBinaryLogger) GetMethodLogger(methodName string) iblog.MethodLogger {
+	var ol, ml iblog.MethodLogger
+	if l.originalLogger != nil {
+		ol = l.originalLogger.GetMethodLogger(methodName)
+	}
+	if l.wrappedLogger == nil {
+		return ol
+	}
+	ml = l.wrappedLogger.GetMethodLogger(methodName)
+	if ml == nil {
+		return ol
+	}
+	return &observabilityBinaryMethodLogger{
+		originalMethodLogger: ol,
+		wrappedMethodLogger:  ml,
+		rpcID:                uuid.NewString(),
+	}
+}
+
+func newObservabilityBinaryLogger(iblogger iblog.Logger) *observabilityBinaryLogger {
+	return &observabilityBinaryLogger{
+		originalLogger: iblogger,
+	}
+}
+
+func compileBinaryLogControlString(config *configpb.ObservabilityConfig) string {
+	if len(config.LogFilters) == 0 {
+		return ""
+	}
+
+	entries := make([]string, 0, len(config.LogFilters)+1)
+	for _, logFilter := range config.LogFilters {
+		// With undefined HeaderBytes or MessageBytes, the intended behavior is
+		// logging zero payload. This detail is different than binary logging.
+		entries = append(entries, fmt.Sprintf("%v{h:%v;m:%v}", logFilter.Pattern, logFilter.HeaderBytes, logFilter.MessageBytes))
```

Comment:
Do we prefer to do it this way, now that we have the option?  Converting our filters into some other config string seems worse to me than making grpc's binlogging log everything and implementing filtering here.

Reply:
@menghanl, @lidizheng and I discussed and he will look into something more like this to avoid converting to a string & back:

```go
type LoggerConfig struct {
	methods []MethodConfig
}

type MethodConfig struct {
	pattern string
	headerLimit int;
	messageLimit int;
}

func NewLoggerWithConfig(lc LoggerConfig) Logger {
	l := &logger{}
	for method := lc.methods {
		l.setServiceMethodLogger(...) // as needed
	}
	return logger
}
```

WDYT?

Reply:
Yeah, SG. And `not`?

Reply:
You mean to control the blacklist?  Yeah I think we'll need something like that, too.

Reply:
And should this list be processed in order (just like path matcher in xds)? (1)
Or it's like a list of key-value pairs that will be turned back into a map (like what's currently done in binary logging)? (2)

I guess 2 is the only option if we want to keep the current binary logging behavior.

Reply:
That's an interesting question.  @lidizheng is it defined what happens if the same / an overlapping filter is specified in the config proto?  E.g. consider:

```
filters: [
  {service: "Foo", method: "*", header: 20bytes},
  {service: "Foo", method: "Bar", header: 80bytes},
]
```

Does `Foo/Bar` take precedence over `Foo/*` because it is more specific?  Or does `Foo/*` take precedence since it comes first in the list?